packages/alfa-aria/src/feature.ts

```diff
@@ -495,7 +458,47 @@ const Features: Features = {
           yield Attribute.of("aria-placeholder", value);
         }
       },
-      nameFromLabel
+      (element, device, state) => {
+        if (
+          test<Element>(
+            hasInputType("text", "password", "search", "tel", "url"),
+            element
+          )
+        ) {
+          return Name.fromSteps(
+            () => nameFromLabel(element, device, state),
+            () => nameFromAttribute(element, "title"),
+            () => nameFromAttribute(element, "placeholder")
+          );
+        }
+
+        if (test<Element>(hasInputType("button"), element)) {
+          return nameFromAttribute(element, "value");
+        }
+
+        if (test<Element>(hasInputType("submit"), element)) {
+          return Name.fromSteps(
+            () => nameFromAttribute(element, "value"),
+            () => Branched.of(Option.of(Name.of("Submit")))
+          );
+        }
+
```

Comment:
I'm not a huge fan of replacing the `switch` with a chain of `if` that I find harder to mentally parse and understand (+lot of repeated boilerplate).
I do agree that moving the input type computation out of this is good, though…

Maybe have a `inputType(element)` helper that can be used both here (`switch(inputType(element))` and in `hasInputType` and does the work?

Reply:
The only "boilerplate" I'm not too keen on having to repeat is the type annotation in `test()`, but that's caused by an underlying issue due to the conflation of predicates and type guards. I'm progressing on a way to solve this though. I do personally prefer the combination of `test()` and a predicate as predicates are much more expressive than `switch` statements. I'd therefore rather work on patterns for making this even easier to express using predicates than fit it into something that we can express using `switch`.

Reply:
It might look like this:

```ts
const role = match(element)
  .when(hasInputType("button"), () => Role.of("button"))
  .when(hasInputType("checkbox"), () => Role.of("checkbox"))
  .getOrElse(() => Role.of("textbox"));
```

That looks surprisingly close to something `Option<T>`-like 🤔

Reply:
Seeing how `Either<L, R>` is already right biased I may be able to tack this stuff directly onto it 🤔

Reply:
@Jym77 Let me know what you think of the above. While it isn't something I'll be introducing in this pull request as it depends on some underlying changes to the `Predicate<T, U extends T>` type which I need to split in two, `Predicate<T>` and `Refinement<T, U extends T>`, it's something I can start working on separately.

Reply:
Oh, another _really_ nice thing about the selective functor pattern used above: There's no implicit fall through foot gun lying around. That's another reason why I highly dislike the `switch` statement.

Reply:
Yeah, `switch` has always been the poor man's pattern matching…
I say this makes sense.